File: src/lib/algorithms/lamport_clock.ts
export class LamportClock {
    private count: number;

    constructor(initialValue: number = 0) {
        this.count = initialValue;
    }

    // Local event: increment clock
    tick(): number {
        this.count += 1;
        return this.count;
    }

    // Send message: tick and return current count
    send(): number {
        return this.tick();
    }

    // Receive message: clock = max(local_clock, received_timestamp) + 1
    receive(receivedTimestamp: number): number {
        this.count = Math.max(this.count, receivedTimestamp) + 1;
        return this.count;
    }

    get time(): number {
        return this.count;
    }
}

File: algorithms/lamport_clock.go
package algorithms

import "sync"

// LamportClock tracks logical time for distributed event ordering.
type LamportClock struct {
	mu     sync.Mutex
	clock  int
	nodeID string
}

func NewLamportClock(nodeID string) *LamportClock {
	return &LamportClock{
		nodeID: nodeID,
	}
}

// Tick is used for local/internal events.
func (lc *LamportClock) Tick() int {
	lc.mu.Lock()
	defer lc.mu.Unlock()

	lc.clock++
	return lc.clock
}

// SendEvent increments logical time before a send.
func (lc *LamportClock) SendEvent() int {
	return lc.Tick()
}

// ReceiveEvent applies Lamport receive rule:
// clock = max(local_clock, received_timestamp) + 1
func (lc *LamportClock) ReceiveEvent(receivedTimestamp int) int {
	lc.mu.Lock()
	defer lc.mu.Unlock()

	if receivedTimestamp > lc.clock {
		lc.clock = receivedTimestamp
	}
	lc.clock++
	return lc.clock
}

func (lc *LamportClock) GetTime() int {
	lc.mu.Lock()
	defer lc.mu.Unlock()

	return lc.clock
}
